content/blog/2022/2022-11-21-building-csv-importer-with-nodejs.md

```diff
@@ -0,0 +1,953 @@
+    },
+    updatedAt: {
+      type: Sequelize.STRING
+    },
+    avatar: {
+      type: Sequelize.STRING
+    }
+  });
+
+export default Employee;
+```
+
+As we can see, we have an attribute as `id` which will be the foreign key for the table. We also have an attribute as `managedBy` which denotes the id of the manager who is managing this employee. We need to mark this attribute as a foreign key. In Sequelize, we can define  `references` and map the reference to the other model and key for foreign key definition.
+
+Next, we need to define ORM mapping for the parent-child relationship. We will have a one-to-many relationship which means one manager can have multiple employees under that person. Sequelize provides 4 types of associations that should be combined to create ORM mappings for *One-To-One*, *One-To-Many* and *Many-To-Many*:
+
+* `hasOne`
+* `belongsTo`
+* `hasMany`
+* `belongsToMany`
+
+In our data model, we have to use the combination of `hasMany` and `belongsTo`:
+
+```javascript
+Employee.hasMany(Employee, {
+    as: 'children',
+    foreignKey: 'managedBy',
+    sourceKey: 'id',
+    useJunctionTable: false
+  });
+Employee.belongsTo(Employee, {
+    foreignKey: "managedBy",
+    targetKey: "id",
+  });
+```
+
+Next, we need to define a database initiator code to initiate and create a table using the above model and association defined. We will create a `database` folder and add our logic:
+
+```javascript
+import Sequelize from 'sequelize';
+import dbConfig from '../config/db.config.js';
+
+export const sequelize =  new Sequelize(dbConfig.DB, dbConfig.USER, dbConfig.PASSWORD, {
+    host: dbConfig.HOST,
+    dialect: dbConfig.dialect,
+    pool: dbConfig.pool,
+    logging: console.log
+  }
+);
+
+sequelize.authenticate()
+  .then(() => {
+    console.log('Connection has been established successfully.');
+    console.log('Creating tables ===================');
+    sequelize.sync().then(() => {
+        console.log('=============== Tables created per model');
+    })
+    .catch(err => {
+        console.error('Unable to create tables:', err);    
+    })
+  })
+  .catch(err => {
+    console.error('Unable to connect to the database:', err);
+});
+```
+
+This will connect to PostgreSQL and `sync` all the tables as per the models defined.
+
+## Building a Middleware to Temporarily Store a CSV File Before Uploading
+
+As mentioned earlier, we are using *multer* as a *body parsing middleware* that handles content type `multipart/form-data` which is primarily used for uploading files. That means it parses the raw HTTP request data which are primarily used for file upload and makes it more accessible by storing it somewhere for further processing. Without multer, we would have to parse the raw data ourselves to access the file.
+
+So let’s define middleware by creating a `middleware` folder and adding our logic inside:
+
+```javascript
+import fs from 'fs';
+import multer from 'multer';
+
+const storage = multer.diskStorage({
```

Comment:
Please add a warning callout somewhere that this code is storing the uploaded file on the disk and that it would be more resource efficient and scalable to stream the CSV data through to the database (which requires some more complicated code).

Reply:
Usually for uploading data, either the files are stored in disk or in memory until the whole data is read out of it. I feel its not required to add a warning as such.